home-assistant: init at 0.62.1

dotlambda opened the pull request:
closes #33675 

Regarding `--skip-pip`:
When I disable that option, not even the frontend component can be loaded because home-assistant isn't able to import pip. I have no idea why.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @FRidh @f-breidenstein @flokli

flokli committed:
astral: init at 1.4

fleaz committed:
Add f-breidenstein as maintainer

Commit:
home-assistant: init at 0.62.1

Commit:
home-assistant: add NixOS module

dotlambda commented:
I have modified the derivation to use an older version of `yarl`, since `yarl-1.0.0` removed the `quote` and `unquote` functions used by home-assistant: https://github.com/aio-libs/yarl/blob/master/CHANGES.rst#100-2018-01-15.

dotlambda commented:
@FRidh Could you please have a look at whether I implemented `extraPackages` exactly as you intended it to be?

FRidh commented on pkgs/development/python-modules/astral/default.nix:
```diff
@@ -0,0 +1,27 @@
+buildPythonPackage rec {
+  pname = "astral";
+  version = "1.4";
+  name  = "${pname}-${version}";
```
no name

FRidh commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,55 @@
+  python = python3.override {
+    packageOverrides = self: super: {
+      yarl = super.yarl.overridePythonAttrs (oldAttrs: rec {
+        name = "${oldAttrs.pname}-${version}";
```
if yarl is without `name`, then `name` can be dropped here as well.

FRidh commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,55 @@
+        };
+      });
+      aiohttp = super.aiohttp.overridePythonAttrs (oldAttrs: rec {
+        name = "${oldAttrs.pname}-${version}";
```
same here

dotlambda replied:
`aiohttp` still had a name attribute. I've removed it now.

FRidh commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,55 @@
+      });
+    };
+  };
+  hass-frontend = python.pkgs.callPackage ./frontend.nix { };
```
having this inside `packageOverrides` and then `self.callPackage` is I think a bit nicer, but it is not necessary.

dotlambda replied:
It also has the advantage of being able to modify the frontend package.

FRidh commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,55 @@
+    requests pyyaml pytz pip jinja2 voluptuous typing aiohttp yarl async-timeout chardet astral certifi
+    # From running hass
+    hass-frontend sqlalchemy netdisco xmltodict user-agents
+  ] ++ extraPackages python.pkgs;
```
I suggest having `extraPackages python.pkgs` either in the `let` expression assigned to something, or have parentheses here around it, just for clarity. I don't know what is better. What do you think?

FRidh commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,55 @@
+
+  makeWrapperArgs = [ "--add-flags --skip-pip" ];
+
+  doCheck = false;
```
comment with reason?

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11954,6 +11954,8 @@ with pkgs;
 
   hiawatha = callPackage ../servers/http/hiawatha {};
 
+  home-assistant = python3Packages.callPackage ../servers/home-assistant { };
```
This should simply be `callPackage` because we do not have any Python packages to supply as arguments.

dotlambda replied:
That's true, but it will hopefully change in the future when home-assistant allows newer versions of the dependencies :)

FRidh replied:
If we have packages from `python-packages.nix` and packages defined locally (read: the referred file) then we need to update the Python package set with `packageOverrides`. There will not be any use for defining Python packages as parameters, and as such no case for using `pythonPackages.callPackage`.

dotlambda replied:
I know, but I hope that home-assistant will just work with the standard python package set in the future.

FRidh replied:
Currently the not overridden Python package set (see arguments `python` and `isPy3k`) is mixed with the overridden set. That should not happen. 

> I know, but I hope that home-assistant will just work with the standard python package set in the future.

Very unlikely. We have a lot of Python applications, and for the larger ones we typically need to provide overrides.

FRidh reviewed (changes requested):
Looks good. I haven't reviewed the module though.

FRidh commented:
> Regarding `--skip-pip`:
> When I disable that option, not even the frontend component can be loaded because home-assistant isn't able to import pip. I have no idea why.

Provide `pip` conditionally?

> Could you please have a look at whether I implemented extraPackages exactly as you intended it to be?

Exactly as intended! :+1:

Commit:
pythonPackages.aiohttp: remove name attribute

dotlambda commented:
Providing `pip` conditionally won't work because home-assistant will still check whether it is running inside a virtualenv. I've previously tried removing https://github.com/home-assistant/home-assistant/blob/master/homeassistant/util/package.py#L10 but some home-assistant component seems to import `running_under_virtualenv`.

dotlambda commented:
It would be nice if someone else could also test the NixOS module. The only problem I've noticed so far that the map doesn't show up on http://localhost:8123/map.

flokli commented:
It might still take me one or two days to give this a spin, but could you provide a nixos test for it, and maybe add a way to configure it using the module directly too, using `toYAML`?

dotlambda commented:
Do you have an idea of what exactly to test? I haven't yet worked with home-assistant that much.

flokli commented:
Maybe simply asserting something is listening on the web interface port and replying with a 2xx code?

dotlambda commented:
@flokli I have added a `config` option that allows specifying the `configuration.yaml` as a Nix attribute set. However, I wasn't sure how to deal with the case of an existing `configuration.yaml` file. I decided to simply delete the old one in this case.
See https://github.com/dotlambda/nixpkgs/blob/a92fddf2cb591d9047182c7d1979a08e4ac8f105/nixos/modules/services/misc/home-assistant.nix#L56.

Commit:
home-assistant: add NixOS test

dotlambda commented:
Finally, a NixOS test has been added. It checks whether
- the `config` option works as expected and
- web interface and API are accessible.

dotlambda commented:
Btw, should I add the test to `nixos/release.nix`?

dotlambda commented:
And finally, I think the package should actually go into `python-packages.nix` since it also provides a Python API: https://dev-docs.home-assistant.io/en/dev/

Commit:
pythonPackages.pytest-aiohttp: init at 0.3.0

dotlambda commented:
I managed to enable the tests included with home-assistant. However, I had to disable most component tests for obvious reasons.
As I see it, this is now ready to be merged.
Did anybody have a thorough look at the NixOS module and test? And should the test be included `nixos/release.nix`?

dotlambda commented:
/cc @peterhoeg @schneefux who have previously worked on #19508

Mounium commented:
I've been using the service for some time now and there doesn't seem to be any problem with the base, and if I add a new module to the configuration file I get a clear error message about what is missing, and so far I could get the modules working I tried (dark sky by creating a package for python-forecastio and webostv by creating one for pylgtv). Thank you for the great work! :)

A question, where should modules/extraPackages go? Simply inside python-modules or should there be a separate folder inside home-assistant's? I think most of these python packages are used exclusively for home-assistant and in the long run a virtualenv sort of solution would be ideal.

dotlambda commented:
Thanks a lot for testing @Mounium! I'm glad you didn't encounter any problems.

Maybe we should start creating a mapping `"component" -> [ python packages ]` or generate it from [requirements_all.txt](https://github.com/home-assistant/home-assistant/blob/master/requirements_all.txt). This should only include the components for which all dependencies are packaged in Nixpkgs. This way, we could automatically generate a list of dependencies from `services.home-assistant.config` and use this list as `extraPackages`.
And then, the modules could go into `python-packages.nix` just as normal.
I personally wouldn't want hass to pull in dependencies dynamically, because this is completely against Nix's principles, doesn't go well with rollbacks etc.

For the time being, I think it would be nice to mention tested (whether working or not) components in this thread, together with their dependencies. But feel free to write above-mentioned script for generating the mapping :)

peterhoeg commented:
Are you guys aware of https://github.com/garbas/nixpkgs-python/tree/master/homeassistant ?

dotlambda commented:
Yes, I am, but this
- is not maintained
- uses pypy2nix and can therfore not be included in Nixpkgs
- doesn't provide a NixOS module (it's possible to use it as the `package` for my module)

dotlambda commented:
I have written a small script that looks up the python packages needed for all components using `nix-env`: https://gist.github.com/dotlambda/83e72cd47d797f6c25c442d6d61bdee7
I still need to generate a nix attribute set using that information.

peterhoeg commented on nixos/modules/services/misc/home-assistant.nix:
```diff
@@ -0,0 +1,91 @@
+        ln -s ${configFile} ${cfg.configDir}/configuration.yaml
+      '';
+      serviceConfig = {
+        PermissionsStartOnly = true; # preStart must be run as root
```
Instead of setting ```PermissionsStartOnly``` and then having to manually ```mkdir``` and ```chown``` the directory, you could use one of the following options:

a) A ```tmpfiles.d``` fragment via ```systemd.tmpfiles.rules```
b) Set it as the home directory for the user and then have it created together with the user

While there is of course nothing technically wrong about your approach it's just nicer to use the declarative tools we have available.

And thanks for packaging this!!!

dotlambda replied:
I went for option b)

dotlambda replied:
Do you have any other comments regarding the module @peterhoeg? Or do you think it can be merged as is?

peterhoeg replied:
The module looks very nice - and well done on the tests!!

Couple of smaller issues:

1. Please set ```serviceConfig.Restart = "on-failure";``` as we probably want it to restart if it crashes.

2. Have you looked at all at locking down the service even further with ```ProtectSystem```, ```PrivateTmp```, ```ReadOnlyDirectories``` and so on?

dotlambda replied:
1. Do we really want this?
I know Arch does it: https://aur.archlinux.org/cgit/aur.git/tree/home-assistant.service?h=home-assistant.
However, the official instructions don't have that: https://home-assistant.io/docs/autostart/systemd/.
What's the point of restarting if it fails? From my knowledge, home-assistant should only fail if there's something wrong with the config. Or do you have experience that suggests otherwise?
2. `PrivateTmp` can't be used because hass stores information on state and history inside `configDir`. `ReadOnlyDIrectories` is not possible for the same reason.
I would however implement `ProtectSystem=strict` and `ReadWritePaths=${configDir}`. Do you think that's an adequate choice?

peterhoeg replied:
Is ```TMP``` inside ```configDir```?

dotlambda replied:
I thought you wanted to put `configDir` inside `/tmp`. Then it wouldn't make sense do delete `/tmp` at every restart. :)
With the current restrictions in place, does `PrivateTmp` have an additional benefit? It could even hinder some use cases, e.g. a certain [file sensor](https://home-assistant.io/components/sensor.file/).

peterhoeg replied:
> I thought you wanted to put configDir inside /tmp. 

If home-assistant needs to persist data, it should do so under /var/lib as you already default it to.

> Then it wouldn't make sense do delete /tmp at every restart. :)

Well, technically PrivateTmp doesn't delete - it simply mounts a new directory every time it runs. But the effect is the same.

> With the current restrictions in place, does PrivateTmp have an additional benefit? It could even hinder some use cases, e.g. a certain file sensor.

If you need to access hass's /tmp from the outside, then it will not work. In terms of benefits, it's all about creating a clean, consistent environment in which to run to minimize the risk of something from the outside messing with it or broken/malicious code messing with things outside.

dotlambda replied:
Okay, I implemented it. I however had to move the test's `configDir` out of `/tmp`.

dotlambda commented:
@GrahamcOfBorg test home-assistant

GrahamcOfBorg reviewed (commented):
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
syncing
hass: running command: sync
hass# [   44.534756] hass[718]: INFO:homeassistant.components.http:Serving /api/ to 127.0.0.1 (auth: True)
hass: exit status 0
test script finished in 45.67s
cleaning up
killing hass (pid 593)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/tmp/nix-build-vm-test-run-home-assistant.drv-0/vde1.ctl': Directory not empty
/nix/store/3h14vh4fjwxfnjhpy08xwdf12dqnas49-vm-test-run-home-assistant
```
</p></details>

dotlambda commented:
@GrahamcOfBorg build home-assistant
@GrahamcOfBorg test home-assistant

@FRidh I believe this is ready for merging.
Remaining ideas like automatically adding dependencies according to the `config` option can be dealt with in further PRs. This way, other people can also contribute.

GrahamcOfBorg reviewed (commented):
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/bs6hd9svzi4w50d97pcp7jqcgnwh9lly-homeassistant-0.62.0
```
</p></details>

GrahamcOfBorg reviewed (commented):
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
hass: exit status 1
syncing
hass: running command: sync
hass: exit status 0
test script finished in 45.61s
cleaning up
killing hass (pid 593)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/tmp/nix-build-vm-test-run-home-assistant.drv-0/vde1.ctl': Directory not empty
/nix/store/j6f4gfhgyq2axydxln5d1pwyj2kk6gn7-vm-test-run-home-assistant
```
</p></details>

GrahamcOfBorg reviewed (commented):
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
tests/components/test_init.py ............                               [ 72%]
tests/components/test_introduction.py .                                  [ 72%]
tests/components/test_logger.py ....                                     [ 75%]
tests/components/test_script.py .....                                    [ 78%]
tests/components/test_shell_command.py ......                            [ 82%]
tests/components/test_system_log.py ..............                       [ 91%]
tests/components/test_websocket_api.py ..............                    [100%]

========================= 159 passed in 30.74 seconds ==========================
/nix/store/42ib64df8rs2yvl788clfz0zbsnawz0l-homeassistant-0.62.0
```
</p></details>

dotlambda commented:
Removed from `python-packages.nix` because this should be a *consistent* set of packages.
Also updated to 0.62.1.
@FRidh Is there anything left to do?

@GrahamcOfBorg build home-assistant
@GrahamcOfBorg test home-assistant

GrahamcOfBorg reviewed (commented):
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
tests/components/test_init.py ............                               [ 72%]
tests/components/test_introduction.py .                                  [ 72%]
tests/components/test_logger.py ....                                     [ 75%]
tests/components/test_script.py .....                                    [ 78%]
tests/components/test_shell_command.py ......                            [ 82%]
tests/components/test_system_log.py ..............                       [ 91%]
tests/components/test_websocket_api.py ..............                    [100%]

========================= 159 passed in 10.68 seconds ==========================
/nix/store/h15jk268357qxdif5kfkjklih96641pa-homeassistant-0.62.1
```
</p></details>

GrahamcOfBorg reviewed (commented):
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
hass: exit status 1
syncing
hass: running command: sync
hass: exit status 0
test script finished in 238.19s
cleaning up
killing hass (pid 596)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/tmp/nix-build-vm-test-run-home-assistant.drv-0/vde1.ctl': Directory not empty
/nix/store/2crgxdwq1adi8hl09g52hl30v65gxv34-vm-test-run-home-assistant
```
</p></details>

GrahamcOfBorg reviewed (commented):
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188)

<details><summary>Partial log (click to expand)</summary><p>

```
tests/components/test_init.py ............                               [ 72%]
tests/components/test_introduction.py .                                  [ 72%]
tests/components/test_logger.py ....                                     [ 75%]
tests/components/test_script.py .....                                    [ 78%]
tests/components/test_shell_command.py ......                            [ 82%]
tests/components/test_system_log.py ..............                       [ 91%]
tests/components/test_websocket_api.py ..............                    [100%]

========================= 159 passed in 31.29 seconds ==========================
/nix/store/64s228galv96j14dl4xbnmagqbydizvz-homeassistant-0.62.1
```
</p></details>

FRidh reviewed (commented):
The Python expressions look good to me. Just the small `callPackage` change is needed. I have not reviewed the module.

dotlambda commented:
Now I don't use `python3Packages.callPackage` anymore.

dotlambda commented:
@GrahamcOfBorg build home-assistant
@GrahamcOfBorg test home-assistant

@FRidh Please merge :)
I will then create another PR with a suggestion for computing the necessary component dependencies from the config in the next days.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/h15jk268357qxdif5kfkjklih96641pa-homeassistant-0.62.1
```
</p></details>

FRidh commented:
Thanks!

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188&attempt_id=466fd163-52ee-4501-bd61-8476f39d546d)

<details><summary>Partial log (click to expand)</summary><p>

```
hass: exit status 1
syncing
hass: running command: sync
hass: exit status 0
test script finished in 47.32s
cleaning up
killing hass (pid 593)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/tmp/nix-build-vm-test-run-home-assistant.drv-0/vde1.ctl': Directory not empty
/nix/store/h8v8n6kaj88p98fxg5xdfxk7aqxng93l-vm-test-run-home-assistant
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
tests/components/test_init.py ............                               [ 72%]
tests/components/test_introduction.py .                                  [ 72%]
tests/components/test_logger.py ....                                     [ 75%]
tests/components/test_script.py .....                                    [ 78%]
tests/components/test_shell_command.py ......                            [ 82%]
tests/components/test_system_log.py ..............                       [ 91%]
tests/components/test_websocket_api.py ..............                    [100%]

========================= 159 passed in 31.11 seconds ==========================
/nix/store/64s228galv96j14dl4xbnmagqbydizvz-homeassistant-0.62.1
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34188&attempt_id=none)

<details><summary>Partial log (click to expand)</summary><p>

```
hass: exit status 1
syncing
hass: running command: sync
hass: exit status 0
test script finished in 58.74s
cleaning up
killing hass (pid 627)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/b3hy82vi2xn51jsrp2dd1yj6q9rbnl3g-vm-test-run-home-assistant
```
</p></details>

yesbox commented on pkgs/servers/home-assistant/default.nix:
```diff
@@ -0,0 +1,72 @@
+  pname = "homeassistant";
+  version = "0.62.1";
+
+  diabled = !isPy3k;
```
Forgot to remove this?

dotlambda replied:
You're right. I'll remove it with the next update.
